### ISSUE TITLE
ASP.NET core 2.2 cookie authentication: unexpected growth of cookies number when the user access token is refreshed

### ISSUE THREAD

**Issue by EnricoMassone** [dotnet/aspnetcore#29075]
Hi, 

I'm working on an ASP.NET core 2.2 web application. 

We authenticate users against a third party open id connect authentication provider and we create the authentication cookie for our application by using the ASP.NET core open id connect authentication middleware. Everything works as expected at the cookie authentication level. 

When the user logs in the resulting authentication cookie is splitted in chunks automatically by ASP.NET core, so after the login we are able to inspect the cookies for the application domain in chrome dev tools and we see 3 different cookies (two of them are the actual chunks, there is also a third cookie which, based on my understanding, is used to keep track that there are two physical chunks). 

So far, so good. 

These days I'm investigating a bug which I'm **not** able to reproduce consistently in all of our environment. **As of today, we have observed this bug only for one of our customers.**

Basically if a user logs in and the browser tab is kept open for a long time, after some hours all the AJAX requests fail due to a 400 response from the server and the application becomes unusable. **The 400 status code is due to a request cookie too large error** : we have obser that, after some hours, the number of cookies increases from the initial value of 3 to a value of 10. **So basically, the number of cookie chunks grows unexpectedly until the application is no more usable due to the giant resulting request cookie.**

After some investagation I noticed that the new cookies (I mean the additional unexpected chunks) are set by a specific HTTP request that our application makes to refresh the user access token. There is a page of our application which makes an AJAX call to a backend endpoint which is used to ensure that the user access token (obtained during the OIDC login and saved inside the authentication properties of the logged in user) is not expired. This is done by using the [identity model ASP.NET core library](https://github.com/IdentityModel/IdentityModel.AspNetCore). 

After some investigation in the code of identity model ASP.NET core I've found the following method, which is responsible to save the fresh new user access token inside the user identity, as an authentication property: 

```
public async Task StoreTokenAsync(ClaimsPrincipal user, string accessToken, DateTimeOffset expiration,
            string refreshToken = null, string resource = null)
        {
            var result = await _contextAccessor.HttpContext.AuthenticateAsync();

            if (!result.Succeeded)
            {
                throw new Exception("Can't store tokens. User is anonymous");
            }

            string tokenName = OpenIdConnectParameterNames.AccessToken;
            if (!string.IsNullOrEmpty(resource))
            {
                tokenName += $"::{resource}";
            }

            string expiresName = "expires_at";
            if (!string.IsNullOrEmpty(resource))
            {
                expiresName += $"::{resource}";
            }


            result.Properties.Items[$".Token.{tokenName}"] = accessToken;
            result.Properties.Items[$".Token.{expiresName}"] = expiration.ToString("o", CultureInfo.InvariantCulture);

            if (refreshToken != null)
            {
                result.Properties.UpdateTokenValue(OpenIdConnectParameterNames.RefreshToken, refreshToken);
            }

            await _contextAccessor.HttpContext.SignInAsync(result.Principal, result.Properties);
        }
```

The original code is available [here](https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/19ec3e6f520e5eac6eff9bc6dc24e31e074227b2/src/AccessTokenManagement/AuthenticationSessionUserTokenStore.cs#L82)

Based on my understanding it's the final call to `HttpContext.SignInAsync` which is responsible to set the cookies in the HTTP response (see [here](https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/19ec3e6f520e5eac6eff9bc6dc24e31e074227b2/src/AccessTokenManagement/AuthenticationSessionUserTokenStore.cs#L113)). 
For some reason, at a certain point of the user session, the cookies set in this HTTP response are many more than the previous number of cookies (as explained above, the number of chunks increases from 3 to 10). **Notice that the user logged in is always the same and the only interaction at the level of the user identity is the refreshing of the user access token**. 

Do you have a possible explanation for this unexpected growth ? 

Maybe before calling `HttpContext.SignInAsync` it is necessary to sign out the current user, in order to delete the existing authentication cookies ? 

Any clue / suggestion in order to understand the root cause of the chunks growth is appreciated. 

Thanks !

**Comment by EnricoMassone** [comment]
I have open a similar issue on the IdentityServer.AspNetCore repository. There I have attached some screen shots of the request and response cookies: https://github.com/IdentityModel/IdentityModel.AspNetCore/issues/174

**Comment by EnricoMassone** [comment]
Additional information:

- the AJAX call to the endpoint used to refresh the user access token authenticate itself by sending authentication cookies with the request
- the called backend endpoint only ensures that the user access token is not expired or about to expire. That's all. This basically means that, at the level of the user identity, either nothing changes or only the user access token changes. 

This is the implmentation of the called action method: 

  ```
  [ApiExplorerSettings(IgnoreApi = true)]
  [Route("boapi/refreshAccessToken")]
  [Authorize]
  public class RefreshAccessTokenController : NonCachableController
  {
    [HttpPost("")]
    public async Task<IActionResult> RefreshAccessToken()
    {
      await HttpContext.GetUserAccessTokenAsync();
      return Ok();
    }
  }
```
The extension method `GetUserAccessTokenAsync` is provided by version `2.0.0` of [IdentityModel.AspNetCore package](https://github.com/IdentityModel/IdentityModel.AspNetCore).

This is a screnn shot of the request and response cookies for an environement which works as expected. Notice that the number of cookies in both the request and response **is the same**, so there is no growth of the number of cookies.  

![test-environment](https://user-images.githubusercontent.com/29275408/103802723-dcae9700-504f-11eb-80a9-9af57b55227f.jpg)

This is a screen shot taken from the not working environment. Notice that, in this case, the number of response cookies is much greater than the number of request cookies. This is the described growth of the number of authentication cookies. 

![offending_environment](https://user-images.githubusercontent.com/29275408/103802824-0962ae80-5050-11eb-80d9-c08dbe2d16b7.jpg)

**Comment by Tratcher** [comment]
Note 2.2 is out of support. The currently supported versions are 2.1, 3.1, and 5.0.

> Maybe before calling HttpContext.SignInAsync it is necessary to sign out the current user, in order to delete the existing authentication cookies ?

The new cookies will overwrite the old ones so there's no need to signout to delete the old cookies.

This sounds like an issue with the refresh logic in IdentityModel, where it provides much more user data during the refresh. The best way to figure that out is to reproduce it under the debugger and inspect StoreTokenAsync's modifications to the Properties collection. If you can't repro under the debugger then you could use the CookieAuthenticationOptions.Events to intercept the SignIn and add logs for the Properties.

**Comment by EnricoMassone** [comment]
At the moment I've patched the code by using client access tokens instead of users access tokens. This is strictly related to how the `IdentityModel.AspNetCore` library works, but to keep it simple and summarize I've changed my implementation in order to workaround this issue and quickly deploy a working solution to the customer.

Unfortunately the only environments where I've been able to reproduce the described cookie growth issue are both the production and staging environments of the client, so my investingation is currently stuck. I would like to further investigate in ordet to understand what happened, but in order to do so I need to be allowed by the customer.  

At the same time we are about to migrate to ASP.NET core 3.1, which could be a big game changer for the problem we encountered (the lastest available version of `IdentityModel.AspNetCore`  has dropped any compatibility with ASP.NET core 2.2).

I'll comment here in case I'll have the chance to further investigate this problem.
